lib/cluster/index.ts

```diff
@@ -0,0 +1,646 @@
+        this.connect().catch(function (err) {
+          debug('Got error %s when reconnecting. Ignoring...', err)
+        })
+      }.bind(this), retryDelay)
+    } else {
+      this.setStatus('end')
+      this.flushQueue(new Error('None of startup nodes is available'))
+    }
+  }
+
+  /**
+   * Disconnect from every node in the cluster.
+   *
+   * @param {boolean} [reconnect=false]
+   * @memberof Cluster
+   */
+  public disconnect(reconnect: boolean = false) {
+    const status = this.status
+    this.setStatus('disconnecting')
+
+    if (!reconnect) {
+      this.manuallyClosing = true
+    }
+    if (this.reconnectTimeout) {
+      clearTimeout(this.reconnectTimeout)
+      this.reconnectTimeout = null
+      debug('Canceled reconnecting attempts')
+    }
+    if (this.slotsTimer) {
+      clearInterval(this.slotsTimer)
+      this.slotsTimer = null
+    }
+
+    this.subscriber.stop()
+    if (status === 'wait') {
+      this.setStatus('close')
+      this.handleCloseEvent()
+    } else {
+      this.connectionPool.reset([])
+    }
+  }
+
+  /**
+   * Quit the cluster gracefully.
+   *
+   * @param {CallbackFunction<'OK'>} [callback]
+   * @returns {Promise<'OK'>}
+   * @memberof Cluster
+   */
+  public quit(callback?: CallbackFunction<'OK'>): Promise<'OK'> {
+    const status = this.status
+    this.setStatus('disconnecting')
+
+    this.manuallyClosing = true
+
+    if (this.reconnectTimeout) {
+      clearTimeout(this.reconnectTimeout)
+      this.reconnectTimeout = null
+    }
+    if (this.slotsTimer) {
+      clearInterval(this.slotsTimer)
+      this.slotsTimer = null
+    }
+
+    const Promise = PromiseContainer.get()
+    if (status === 'wait') {
+      const ret = asCallback(Promise.resolve('OK'), callback)
+
+      // use setImmediate to make sure "close" event
+      // being emitted after quit() is returned
+      setImmediate(function () {
+        this.setStatus('close')
+        this.handleCloseEvent()
+      }.bind(this))
+
+      return ret
+    }
+    return asCallback(
+      Promise.all(this.nodes().map(function (node) {
+        return node.quit()
+      })).then(function () {
+        return 'OK'
+      }),
+      callback
+    )
+  }
+
+  /**
+   * Get nodes with the specified role
+   *
+   * @param {NodeRole} [role='all']
+   * @returns {any[]}
+   * @memberof Cluster
+   */
+  public nodes(role: NodeRole = 'all'): any[] {
+    if (role !== 'all' && role !== 'master' && role !== 'slave') {
+      throw new Error('Invalid role "' + role + '". Expected "all", "master" or "slave"')
+    }
+    return this.connectionPool.getNodes(role)
+  }
+
+  /**
+   * Change cluster instance's status
+   *
+   * @private
+   * @param {ClusterStatus} status
+   * @memberof Cluster
+   */
+  private setStatus(status: ClusterStatus): void {
+    debug('status: %s -> %s', this.status || '[empty]', status)
+    this.status = status
+    process.nextTick(this.emit.bind(this, status))
```

Comment:
this.emit.bind could be saved as a reference, ie boundEmit
and then we can `process.nextTick(this.boundEmit, status)` - that should be a bit faster